plugins/dell-dock/fu-plugin-dell-dock.c

```diff
@@ -184,6 +184,25 @@ fu_plugin_device_registered (FuPlugin *plugin, FuDevice *device)
 		return;
 	}
 
+	/* steal needs_activation if EC has pending update */
+	if (g_strcmp0(fu_device_get_plugin(device), "dell_dock") == 0 &&
+	    FU_IS_DELL_DOCK_USB4(device)) {
+		GPtrArray *devices = fu_plugin_get_devices(plugin);
+		for (guint i = 0; i < devices->len; i++) {
+			FuDevice *device_tmp = g_ptr_array_index(devices, i);
+			if (FU_IS_DELL_DOCK_EC(device_tmp) &&
+			    fu_device_has_flag(device_tmp, FWUPD_DEVICE_FLAG_NEEDS_ACTIVATION)) {
```

Comment:
I'm confused - if the EC has an activation pending that doesn't mean USB4 also has an activation pending.  If you have an update that is EC + package version only, but not USB4 NVM, doesn't that mean this will leave USB4 NVM in needs activation state EVERY SINGLE TIME?

I would think it's better to use history database to set this information, rather than mistakingly taking it from EC needs-activation.

Otherwise this scenario happens:
1. Plug in dock
2. It enumerates.
2. Issue update that is EC and package only
3. Notice that EC is in needs activation.
4. Restart `fwupd.service` or  reboot (don't unplug dock)
5. Notice that USB4 is in needs activation.

Reply:
I think specifically what is missing is is a check that usb4 device already had needs activation flag.  Then you just "remove" the EC's needs-activation to prevent the problem, instead of moving it over every time.

Reply:
> I think specifically what is missing is is a check that usb4 device already had needs activation flag. Then you just "remove" the EC's needs-activation to prevent the problem, instead of moving it over every time.

For running `fwupd` with daemon that is okay as `needs_activation` is tagged to usb4 device. But for `fwupdtool` that is blindsided from `history` doesn't know that usb4 has `needs_activation` flag nor the `update_state`, ended up ignored usb4.

Reply:
Why doesn't it come from history database? That sounds like the bug this is working around to me...

Reply:
Try adding `FU_DEVICE_INTERNAL_FLAG_INHERIT_ACTIVATION` to the device.  If that works - it's way cleaner, and you can just remove the activation flag from EC device when both are set.

Reply:
Adding `FU_DEVICE_INTERNAL_FLAG_INHERIT_ACTIVATION` brings the flag to `usb4` device, thank you. 

Can we shift `fu_engine_device_inherit_history` earlier than `fu_engine_plugin_device_register`? I wanted to address the mutual exclusive problem in `_registered` but the flag comes at the very end.

```
  /* notify all plugins about this new device */
  if (!fu_device_has_flag (device, FWUPD_DEVICE_FLAG_REGISTERED))
	  fu_engine_plugin_device_register (self, device);
  ...
  ...
  
  /* sometimes inherit flags from recent history */
  fu_engine_device_inherit_history (self, device);

```

Reply:
I don't see a problem with swapping the order here.